Login modal

j-alicia-long opened the pull request:
# Status
Done

# What I did
- Updated UI for Login 
     - Created login modal so users aren't forced to be redirected
     - Redirect to browse page with error if login fails
- Linked modal on "Sign In" links on navbar, sidebar, landing page

## Issues addressed
- This closes #132 
- This closes #272 

## Screenshots
| Login modal      | Login Error |
| ----------- | ----------- |
|  ![image](https://user-images.githubusercontent.com/4724192/102069112-534ad280-3dcb-11eb-9da8-3a4800c5776c.png) | ![image](https://user-images.githubusercontent.com/4724192/102069252-855c3480-3dcb-11eb-9a34-fdede6901d73.png) |

julesfll commented on tcf_website/templates/about/logo-tagline.html:
```diff
@@ -0,0 +1,10 @@
+<!-- Logo with name & tagline, import where needed -->
```
Good refactor

julesfll commented on tcf_website/templates/login/login_modal.html:
```diff
@@ -0,0 +1,27 @@
+        </button>
+      </div>
+      <div class="modal-body">
+          <div class="d-flex justify-content-center">
```
Would this be a place to use your `logo-tagline`?

j-alicia-long replied:
I tried but it looks kinda bad in the modal on small screens rip

julesfll replied:
Ok maybe something to look into later
We need to put that refactor to use 😄

julesfll reviewed (commented):
Modal looks good, and this is a good simplification! I do have some suggestions though for usability.
- If you log in incorrectly the first time, you get redirected to `/login/error`. Then if you log in correctly, instead of getting redirected to `Browse`, you get redirected to `/login/error` again (and the message is still there), so it looks you logged in incorrectly again
- You can still log in when you're already logged in, should probably be disabled (or force redirect when logged in to browse)
- If you log in incorrectly from your current page, you lose where you were by being redirected to `login`

What would solve all these issuse would be to eliminate the login page entirely, which might be the best move. The messages communicate a successful/unsuccessful login, so you can just redirect to the current page no matter the outcome.

j-alicia-long commented:
> Modal looks good, and this is a good simplification! I do have some suggestions though for usability.
> 
> * If you log in incorrectly the first time, you get redirected to `/login/error`. Then if you log in correctly, instead of getting redirected to `Browse`, you get redirected to `/login/error` again (and the message is still there), so it looks you logged in incorrectly again
> * You can still log in when you're already logged in, should probably be disabled (or force redirect when logged in to browse)
> * If you log in incorrectly from your current page, you lose where you were by being redirected to `login`
> 
> What would solve all these issuse would be to eliminate the login page entirely, which might be the best move. The messages communicate a successful/unsuccessful login, so you can just redirect to the current page no matter the outcome.

Good point! I'll look into removing the page. Thought my design of it was pretty ugly anyway ngl

j-alicia-long commented:
> Modal looks good, and this is a good simplification! I do have some suggestions though for usability.
> 
> 1. If you log in incorrectly the first time, you get redirected to `/login/error`. Then if you log in correctly, instead of getting redirected to `Browse`, you get redirected to `/login/error` again (and the message is still there), so it looks you logged in incorrectly again
> 2. You can still log in when you're already logged in, should probably be disabled (or force redirect when logged in to browse)
> 3. If you log in incorrectly from your current page, you lose where you were by being redirected to `login`
> 
> What would solve all these issuse would be to eliminate the login page entirely, which might be the best move. The messages communicate a successful/unsuccessful login, so you can just redirect to the current page no matter the outcome.

So removing the login page turned out to be more tricky than anticipated with django auth. I basically replaced the login page in the workflow with the browse page. 
- Fixed the first issue with an if-statement checking the view for including the "next" redirect.
- Fixed the second issue with a force redirect to profile/browse at the `/login` url
- Couldn't fix third issue due to aforementioned trickiness

We might need an overhaul of the auth system in general since its inflexibility has been causing other issues with extra data collection.

alex-shen1 commented on tcf_website/views/auth.py:
```diff
@@ -6,19 +6,22 @@
-
+        messages.success(request, "Logged in successfully!")
+        return redirect('profile')
+    return browse(request)
```
Just curious — when does this return get reached?

j-alicia-long replied:
When the user isn't authenticated and they try to go to `/login`

alex-shen1 commented on tcf_website/views/auth.py:
```diff
@@ -53,4 +56,4 @@ def logout(request):
     auth_logout(request)
     messages.add_message(request, messages.SUCCESS, "Logged out successfully!")
-    return redirect('login')
+    return redirect('browse')
```
What's the difference between `browse(request)` like the others have and this?

j-alicia-long replied:
I can't use `redirect('browse')` in `login` or `login_error` views without causing a `ERR_TOO_MANY_REDIRECTS`. I'm honestly not sure entirely why

j-alicia-long commented:
@julesfll I opened Login Error Redirect Issue #328 to address the third comment!
